### PR TITLE
RTT Programming and Virtual Collaboration

### PR DESCRIPTION
**Pull request by KanishkT123**
This PR adds in:

* A primer on Virtual Collaboration
* Information on how to think about setting up virtual collaboration
* Information on Red Team Testing, a new method for working collaboratively across timezones and virtually

### CODE REVIEW

**Comment by meulta** [comment]
Thanks for the PR, i'll review in the next day or so!

File: agile-development/collaboration/Virtual_Collaboration.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+
+Pair programming works well under the correct circumstances, but it loses some of its charm when executed in a completely virtual setting. The virtual setup still involves two programmers looking at the same screen and talking out their designs, but there are often logistical issues to deal with, including lag, microphone set up issues, workspace and personal considerations, and many other small, individually trivial problems that worsen the experience.
+
+Virtual work patterns need to deviate from the in-person patterns we are accustomed to. Pair programming at its core is based on the following principles:
+
+1. Generating clarity through communication
+2. Producing higher quality through collaboration
+3. Creating ownership through equal contribution
+
+Pair programming is simply one way to achieve these results. We present Red Team Testing (RTT) as a different programming method that takes the same principles but uses some of the advantages that virtual work methods provide.
```
Comment (meulta) [inline agile-development/collaboration/Virtual_Collaboration.md:13]: I would remove the "we present" which makes it a bit too much like an announcement. I would consider changing "different" with "alternate"

File: agile-development/collaboration/Virtual_Collaboration.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+
+## Red Team Testing
+
+Red Team Testing borrows its name from the “Red Team” and “Blue Team” paradigm of penetration testing, and is a collaborative, parallel way of working virtually. In Red Team Testing, two programmers jointly decide on the interface, architecture, and design of the program, and then separate for the implementation phase. One programmer writes tests using the public interface, attempting to perform edge case testing, input validation, and otherwise stress testing the interface. The second programmer is simultaneously writing the implementation which will eventually be tested.
+
+Red Team Testing has the same philosophy as any other Test-Driven Development lifecycle: All implementation is divorced from the interface, and the interface can be tested with no knowledge of the implementation.
```
Comment (meulta) [inline agile-development/collaboration/Virtual_Collaboration.md:19]: Let's use "distinct" or "separated" instead of "divorced"

File: agile-development/collaboration/Virtual_Collaboration.md (hunk 1 above)
Comment (meulta) [reply agile-development/collaboration/Virtual_Collaboration.md:13]: The reason I want to insist on alternate is because I don't think this would be a good fit for ALL teams, it's an option to choose from :)

File: agile-development/collaboration/Virtual_Collaboration.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+
+RTT has many of the same benefits as Pair Programming and Test-Driven development but tries to update them for a virtual setting.
+
+* Code implementation and testing can be done in parallel, over long distances or across time zones, which reduces the overall time taken to finish writing the code.
+
+* RTT maintains the pair programming paradigm, while reducing the need for video communication or constant communication between engineers.
+
+* RTT allows detailed focus on design and engineering alignment before implementing any code, leading to cleaner and simpler interfaces.
+
+* RTT forces testing to be prioritized alongside implementation, instead of having testing follow or be influenced by the implementation of the code.
```
Comment (meulta) [inline agile-development/collaboration/Virtual_Collaboration.md:56]: "encourages"?

**Review by meulta — CHANGES_REQUESTED** [review]

**Comment by meulta** [comment]
Looks great, I added a couple of changes to make if possible :)

File: agile-development/collaboration/Virtual_Collaboration.md (hunk 1 above)
Comment (TessFerrandez) [inline agile-development/collaboration/Virtual_Collaboration.md:13]:
```suggestion
Pair programming is one way to achieve these results. Red Team Testing (RTT) is an alternate programming method that uses the same principles but with some of the advantages that virtual work methods provide.
```

File: agile-development/collaboration/Virtual_Collaboration.md [hunk 4]
```diff
@@ -0,0 +1,66 @@
+# Virtual Collaboration and Pair Programming
+
+Pair programming is the de facto work method that most large engineering organizations use for “hands on keyboard” coding. Two programmers, working synchronously, looking at the same screen and attempting to code and design together, which often results in better and clearer code than either could produce individually.
+
+Pair programming works well under the correct circumstances, but it loses some of its charm when executed in a completely virtual setting. The virtual setup still involves two programmers looking at the same screen and talking out their designs, but there are often logistical issues to deal with, including lag, microphone set up issues, workspace and personal considerations, and many other small, individually trivial problems that worsen the experience.
+
+Virtual work patterns need to deviate from the in-person patterns we are accustomed to. Pair programming at its core is based on the following principles:
```
Comment (TessFerrandez) [inline agile-development/collaboration/Virtual_Collaboration.md:7]:
```suggestion
Virtual work patterns are different from the in-person patterns we are accustomed to. Pair programming at its core is based on the following principles:
```

File: agile-development/collaboration/Virtual_Collaboration.md (hunk 2 above)
Comment (TessFerrandez) [inline agile-development/collaboration/Virtual_Collaboration.md:19]:
```suggestion
Red Team Testing has the same philosophy as any other Test-Driven Development lifecycle: All implementation is separated from the interface, and the interface can be tested with no knowledge of the implementation.
```

File: agile-development/collaboration/Virtual_Collaboration.md [hunk 5]
```diff
@@ -0,0 +1,66 @@
...
+
+Red Team Testing borrows its name from the “Red Team” and “Blue Team” paradigm of penetration testing, and is a collaborative, parallel way of working virtually. In Red Team Testing, two programmers jointly decide on the interface, architecture, and design of the program, and then separate for the implementation phase. One programmer writes tests using the public interface, attempting to perform edge case testing, input validation, and otherwise stress testing the interface. The second programmer is simultaneously writing the implementation which will eventually be tested.
+
+Red Team Testing has the same philosophy as any other Test-Driven Development lifecycle: All implementation is divorced from the interface, and the interface can be tested with no knowledge of the implementation.
+
+![ptt-diagram](PTTdiagram.PNG)
+
+## Steps
+
+1. Design Phase: Both programmers will design the interface together. This includes:
```
Comment (TessFerrandez) [inline agile-development/collaboration/Virtual_Collaboration.md:25]:
```suggestion
1. Design Phase: Both programmers design the interface together. This includes:
```

File: agile-development/collaboration/Virtual_Collaboration.md [hunk 6]
```diff
@@ -0,0 +1,66 @@
...
+    * Method signatures and names
+    * Writing documentation or docstrings for what the methods are intended to do.
+    * Architecture decisions that would influence testing (Factory patterns, etc.)
+
+2. Implementation Phase: The programmers will separate and parallelize work, while continuing to communicate.
```
Comment (TessFerrandez) [inline agile-development/collaboration/Virtual_Collaboration.md:30]:
```suggestion
2. Implementation Phase: The programmers separate and parallelize work, while continuing to communicate.
```

File: agile-development/collaboration/Virtual_Collaboration.md [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+    * Programmer A will design the implementation of the methods, adhering to the previously decided design.
+    * Programmer B will concurrently write tests for the same method signatures, without knowing details of the implementation.
+
+3. Integration & Testing Phase: Both programmers will commit their code and run the tests.
```
Comment (TessFerrandez) [inline agile-development/collaboration/Virtual_Collaboration.md:34]:
```suggestion
3. Integration & Testing Phase: Both programmers commit their code and run the tests.
```

File: agile-development/collaboration/Virtual_Collaboration.md [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+
+RTT also works well when there is complete consensus, or no consensus at all, on what purpose the code serves. Since creating the design jointly and agreeing to implement and test against it are part of the RTT method, RTT forcibly creates clarity through iteration and communication.
+
+## Benefits
+
+RTT has many of the same benefits as Pair Programming and Test-Driven development but tries to update them for a virtual setting.
+
+* Code implementation and testing can be done in parallel, over long distances or across time zones, which reduces the overall time taken to finish writing the code.
+
+* RTT maintains the pair programming paradigm, while reducing the need for video communication or constant communication between engineers.
```
Comment (TessFerrandez) [inline agile-development/collaboration/Virtual_Collaboration.md:52]: Here we use the term engineers - until now the text has consistently referred to them as programmers - I would suggest to change both programmers and engineers to developers throughout the text to be consistent with the language we use in the rest of the playbook

File: agile-development/collaboration/Virtual_Collaboration.md (hunk 3 above)
Comment (TessFerrandez) [inline agile-development/collaboration/Virtual_Collaboration.md:56]:
```suggestion
* RTT encourages testing to be prioritized alongside implementation, instead of having testing follow or be influenced by the implementation of the code.
```

File: agile-development/collaboration/Virtual_Collaboration.md [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+
+* RTT allows detailed focus on design and engineering alignment before implementing any code, leading to cleaner and simpler interfaces.
+
+* RTT forces testing to be prioritized alongside implementation, instead of having testing follow or be influenced by the implementation of the code.
+
+* Documentation is inherently a part of RTT, since both the implementer and the tester need correct, up to date documentation in the implementation phase.
```
Comment (TessFerrandez) [inline agile-development/collaboration/Virtual_Collaboration.md:58]:
```suggestion
* Documentation is inherently a part of RTT, since both the implementer and the tester need correct, up to date documentation, in the implementation phase.
```

**Review by TessFerrandez — CHANGES_REQUESTED** [review]
Nice document - I like this approach.

Added some language suggestions and turned some of the suggestions proposed by @meulta to github suggestions so they are quicker to incorporate

File: agile-development/collaboration/Virtual_Collaboration.md (hunk 1 above)
Comment (KanishkT123) [reply agile-development/collaboration/Virtual_Collaboration.md:13]: Thank you, this helps a lot!

File: agile-development/collaboration/Virtual_Collaboration.md (hunk 8 above)
Comment (KanishkT123) [reply agile-development/collaboration/Virtual_Collaboration.md:52]: Makes sense, we can definitely do that. Thank you!

**Review by meulta — APPROVED** [review]

**Review by TessFerrandez — APPROVED** [review]
